### DISCUSSION TITLE
Header chunk and the number of tracks

### DISCUSSION THREAD

**Discussion by kimyj1503** [FluidSynth/fluidsynth#1291] [Q&A] [answered]
Hello. I recently started studying MIDI messages and Fluidsynth.
After reading some document about the header chunk, I checked the related code on _src/midi/fluid_midi.c_.
I have some trouble understanding some codes and any help would be much appreciated. 

&nbsp;
The structure of the header chunk of the MIDI file.

||chunk ID|length|format type|number of tracks|time division|
|---|---|---|---|---|---|
|size in byte|4|4|2|2|2|
|value|"MThd"|6|0,1, or 2| 1 to 65535|0 or 1 for MSB|

The header chunk of the MIDI file is read through the function [**fluid_midi_file_read_mthd**](https://github.com/FluidSynth/fluidsynth/blob/5ecdc4568e45123216c6888892caad07918ef127/src/midi/fluid_midi.c#L342).
&nbsp;

The below is how I understand the code about the time division, when the MSB equals 0.
```
mf->uses_smpte = 0;
mf->division = ((unsigned)mthd[12] << 8) | ((unsigned)mthd[13] & 0xff);

/*
The lower case alphabets here symbol the binary value of 0 or 1.
Lets say the binary values are as below
mthd[12] = 0abc defg
mthd[13] = hijk lmnp

Then
mf-> division will become 0abc defg hijk lmnp
*/
```
&nbsp;
&nbsp;

But I can't really understand the code for the number of tracks.
**Why is _mthd[10]_ shifted 16 bits instead of 8 bits?**
```
mf->ntracks = (unsigned) mthd[11];
mf->ntracks += (unsigned int)(mthd[10]) << 16;

/*
The lower case alphabets here symbol the binary value of 0 or 1.
Lets say the binary values are as below
mthd[10] = abcd efgh
mthd[11] = ijkl mnpq

Then
mf-> ntracks will NOT become abcd efgh ijkl mnpq
*/
```

I understand that due to the below lines in the [_**fluid_player_add_track**_ function](https://github.com/FluidSynth/fluidsynth/blob/5ecdc4568e45123216c6888892caad07918ef127/src/midi/fluid_midi.c#L1839), **_ntracks_** value larger than 127 won't be used in other part of the code and **_ntracks_** value smaller than 128 will be converted to the unsigned integer as intended.
```
if(player->ntracks < MAX_NUMBER_OF_TRACKS)
    {
        player->track[player->ntracks++] = track;
        return FLUID_OK;
    }
```

Thank you in advance.

**Comment by derselbst** [comment, marked as answer]
> Why is mthd[10] shifted 16 bits instead of 8 bits?

Because it's a bug in 20 year old legacy code. And it doesn't make a difference, because of the `MAX_NUMBER_OF_TRACKS` check you've correctly pointed out. Yet, you're welcome to draft a PR.
